project/src/drat.rs

```diff
@@ -0,0 +1,286 @@
+use aes_gcm_siv::{Aes256GcmSiv, Nonce};
+use hkdf::Hkdf;
+use hmac::{Hmac, Mac};
+use sha2::Sha256;
+use x25519_dalek::{PublicKey, ReusableSecret};
+
+type HmacSha256 = Hmac<Sha256>;
+
+#[derive(Default, PartialEq, Eq, Clone)]
+pub struct SymmKey([u8; 32]);
+
+pub struct State {
+    dhsk_snd: ReusableSecret,
+    dhpk_snd: PublicKey,
+    dhpk_rcv: PublicKey,
+    rt_k: SymmKey,
+    ck_snd: SymmKey,
+    ck_rcv: SymmKey,
+    n_snd: u64,
+    n_rcv: u64,
+    prev_n: u64,
+    mk_skipped: HashMap<(PublicKey, u64), SymmKey>,
+}
+
+pub struct Header {
+    dhpk: PublicKey,
+    prev_n: u64,
+    n: u64,
+}
+
+impl SymmKey {
+    pub fn new(key_bytes: [u8; 32]) -> Self {
+        SymmKey(key_bytes)
+    }
+
+    pub fn first_key_byte(&self) -> u8 {
+        self.0[0]
+    }
+}
+
+impl Display for SymmKey {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "Symmkey({:?})", self.0)
+    }
+}
+
+impl Header {
+    pub const HEADER_LEN: usize = 48;
+    pub const PK_LEN: usize = 32;
+
+    fn new(dhpk: PublicKey, prev_n: u64, n: u64) -> Self {
+        Self { dhpk, prev_n, n }
+    }
+
+    fn as_bytes(&self) -> [u8; Self::HEADER_LEN] {
+        let mut bytes = [0u8; Self::HEADER_LEN];
+        let (dhpk_dest, tail) = bytes.split_at_mut(Self::PK_LEN);
+        let (prev_n_dest, n_dest) = tail.split_at_mut(8);
+        dhpk_dest.copy_from_slice(self.dhpk.as_bytes());
+        prev_n_dest.copy_from_slice(&self.prev_n.to_le_bytes());
+        n_dest.copy_from_slice(&self.n.to_le_bytes());
+        bytes
+    }
+}
+
+impl State {
+    pub const MAX_SKIP: u64 = 100;
+    // helper functions
+
+    /// function to derive new root key and chain key from shared secret and old root key using HKDF
+    fn kdf_root(
+        rt_k: &SymmKey,
+        dhsk_snd: &ReusableSecret,
+        dhpk_rcv: &PublicKey,
+    ) -> (SymmKey, SymmKey) {
+        let dh_out = dhsk_snd.diffie_hellman(dhpk_rcv);
+        let hk = Hkdf::<Sha256>::new(Some(&rt_k.0), dh_out.as_bytes());
+        let mut rk_bytes = [0u8; 32];
+        let mut ck_bytes = [0u8; 32];
+        hk.expand("root-keygen".as_bytes(), &mut rk_bytes).unwrap();
+        hk.expand("chain-keygen".as_bytes(), &mut ck_bytes).unwrap();
+        (SymmKey(rk_bytes), SymmKey(ck_bytes))
+    }
+
+    /// function to hash chain key using HMAC to return new chain key and message key
+    ///
+    /// ```
+    /// use project::drat::{State, SymmKey};
+    ///
+    /// let k = SymmKey::new([0u8; 32]);
+    /// let (ck, mk) = State::kdf_chain(&k);
+    /// println!("{} {}", ck, mk);
+    /// assert_eq!(ck.first_key_byte(), 92u8);
+    /// assert_eq!(mk.first_key_byte(), 235u8);
+    /// ```
+    pub fn kdf_chain(chn_k: &SymmKey) -> (SymmKey, SymmKey) {
+        let mut mac = <HmacSha256 as Mac>::new_from_slice(&chn_k.0).unwrap();
+        mac.update(b"chain");
+        let ck_bytes = mac.finalize_reset().into_bytes();
+        mac.update(b"message");
+        let mk_bytes = mac.finalize().into_bytes();
```

Comment:
Instead of `finalize_reset` you could write the following:
```rust
let mut mac1 = <HmacSha256 as Mac>::new_from_slice(&chn_k.0).unwrap();
let mut mac2 = mac1.clone();

mac1.update(b"chain");
let ck_bytes = mac1.finalize().into_bytes();


mac1.update(b"message");
let ,k_bytes = mac1.finalize().into_bytes();
```
It would allow you to disable the `reset` crate feature, which will make size of `Hmac` a bit smaller.